### PR TITLE
Track the user's relay list

### PR DESCRIPTION
**Pull request by ksedgwic**
This is an alternative approach to #346 

Instead of adding sync to convince rust that the operations are safe we assert they are safe (using `unsafe`) because we are single threaded.

The last commit in the stack is the interesting part ...

### CODE REVIEW

**Comment by ksedgwic** [comment]
rebased on master

File: src/task.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+use nostrdb::{Filter, Ndb, Transaction};
+use uuid::Uuid;
+
+use crate::note::NoteRef;
+use crate::Damus;
+
+pub fn spawn_track_user_relays(damus: &mut Damus) {
+    // This is only safe because we are absolutely single threaded ...
+    let damus_ptr = &mut *damus as *mut Damus;
+    spawn_sendable(async move {
```
Comment (kernelkind) [inline src/task.rs:19]: I think this implementation for adding relays can be simplified into a synchronous implementation that runs every frame in `app::update_damus`.

In the case where the user is actively using the app, the time between frames is going to be so low that it won't matter that we'll have to wait until
the next frame to check ndb for a new relay list. In the case where the user isn't using the app, the update function still runs, about once per second
it seems so if the user is just staring at their relay list without any interaction and an update to their list occurs, they will see it, albeit with
about a second delay.

Do you see a reason that I don't for the increased complexity of this async impl?

**Review by kernelkind — COMMENTED** [review]
TLDR: I think this implementation for adding relays can be simplified into a synchronous implementation that runs every frame in `app::update_damus`

File: src/task.rs (hunk 1 above)
Comment (ksedgwic) [reply src/task.rs:19]: I think polling for activity doesn't scale well to apps with lots of things to poll for.  So it might be fine to poll for user relay lists, but when we add many many other things it will be less great

**Review by jb55 — CHANGES_REQUESTED** [review]
I merged a few of these commits into master, up until the point that needs changes

```
commit 702f8ffca390eb77977556754daceab1e351f9de (HEAD -> master, github/master, test)
Merge: 877a30d2f666 570d64c3cdca
Author: William Casarin <jb55@jb55.com>
Date:   Mon Nov 11 11:16:48 2024 -0800

    Merge a few relay fixes from ken
    
    Pick a few fixes from #380
    
    Link: https://github.com/damus-io/notedeck/pull/380
    
    Ken Sedgwick (5):
          build: Cargo.lock updates to mitigate num_enum_derive problem
          add .rustfmt.toml to specify edition
          Fix parsing of subscription id for RelayMessage::Event
          Skip adding relays that are already in the pool
          canonicalize relay urls to avoid false duplicates

```

File: src/task.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+fn user_relay_filter(damus: &mut Damus) -> Filter {
+    let account = damus
+        .accounts
+        .get_selected_account()
+        .as_ref()
+        .map(|a| a.pubkey.bytes())
+        .expect("selected account");
+
+    // NIP-65
+    Filter::new()
```
[the remark this answers is not in the document]
Comment (ksedgwic) [reply src/task.rs:76]: I'm not sure what you mean here.

Do you mean something in the subscription?

Or discrimination in the handling? We definitely want to discriminate read and write relays but that all needs to be added at the relay management layer as well?

File: src/task.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        .expect("query results");
+    results
+        .iter()
+        .map(|qr| NoteRef::new(qr.note_key, qr.note.created_at()))
+        .filter_map(|nr| ndb.get_note_by_key(txn, nr.key).ok())
+        .map(|n| n.json().unwrap())
+        .collect()
+}
+
+fn handle_nip65_relays(ndb: &Ndb, txn: &Transaction, nks: &Vec<NoteKey>) -> Vec<String> {
```
[the remark this answers is not in the document]
Comment (ksedgwic) [reply src/task.rs:97]: fixed, apologies

File: src/task.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                .iter()
+                .filter_map(|ti| ti.get_unchecked(1).variant().str())
+                .map(|s| s.to_string())
+        })
+        .collect()
+}
+
+fn set_advertised_relays(pool: &mut RelayPool, relays: Vec<String>) {
+    let wakeup = move || {
+        // FIXME - how do we repaint?
```
[the remark this answers is not in the document]
Comment (ksedgwic) [reply src/task.rs:111]: There is not a `ctx` around in this code.  It appears to repaint in the next frame anyway?  I'm not sure why this is necessary

File: src/task.rs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    spawn_sendable(async move {
+        let damus = unsafe { &mut *damus_ptr };
+        track_user_relays(damus).await;
+    });
+}
+
+pub async fn track_user_relays(damus: &mut Damus) {
+    debug!("track_user_relays starting");
+
+    let filter = user_relay_filter(damus);
```
[the remark this answers is not in the document]
Comment (ksedgwic) [reply src/task.rs:28]: agreed
